src/extensions/metadata.rs

```diff
@@ -0,0 +1,179 @@
+//! Adds support for the IMAP METADATA extension specificed in [RFC
+//! 5464](https://tools.ietf.org/html/rfc5464).
+
+use crate::client::*;
+use crate::error::{Error, ParseError, Result};
+use crate::parse::handle_unilateral;
+use crate::types::*;
+use imap_proto::types::{MailboxDatum, Metadata, Response};
+use std::io::{Read, Write};
+use std::sync::mpsc;
+
+trait CmdListItemFormat {
+    fn format_as_cmd_list_item(&self) -> String;
+}
+
+impl CmdListItemFormat for Metadata {
+    fn format_as_cmd_list_item(&self) -> String {
+        format!(
+            "{} {}",
+            validate_str(self.entry.as_str()).unwrap(),
+            self.value
+                .as_ref()
+                .map(|v| validate_str(v.as_str()).unwrap())
+                .unwrap_or("NIL".to_string())
+        )
+    }
+}
+
+/// Represents variants of DEPTH parameters for GETMETADATA command.
+///  "0" - no entries below the specified entry are returned
+///  "1" - only entries immediately below the specified entry are returned
+///  "infinity" -  all entries below the specified entry are returned
+/// See [RFC 5464, section 4.2.2](https://tools.ietf.org/html/rfc5464#section-4.2.2)
```

Comment:
This doc comment won't render nicely, since in Markdown linebreaks aren't significant. We'll probably want to make this a list (start each item with a `-`), and have an empty line above and below the list to make sure we get paragraph breaks. I also think it'd be great to include some concrete examples here of what paths would or would not be matched with different parameters (the same way the RFC does).